Libraries/LibCore/EventLoopImplementationWindows.cpp

```diff
@@ -0,0 +1,241 @@
+/*
+ * Copyright (c) 2023, Andreas Kling <andreas@ladybird.org>
+ * Copyright (c) 2024, stasoid <stasoid@yahoo.com>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include <LibCore/EventLoopImplementationWindows.h>
+#include <LibCore/Notifier.h>
+#include <LibCore/ThreadEventQueue.h>
+#include <WinSock2.h>
+#include <io.h>
+
+struct Handle {
+    HANDLE handle = NULL;
+
+    explicit Handle(HANDLE h = NULL)
+        : handle(h)
+    {
+    }
+    Handle(Handle&& h)
+    {
+        handle = h.handle;
+        h.handle = NULL;
+    }
+    void operator=(Handle&& h)
+    {
+        VERIFY(!handle);
+        handle = h.handle;
+        h.handle = NULL;
+    }
+    ~Handle()
+    {
+        if (handle)
+            CloseHandle(handle);
+    }
+
+    bool operator==(Handle const& h) const { return handle == h.handle; }
+    bool operator==(HANDLE h) const { return handle == h; }
+};
+
+template<>
+struct Traits<Handle> : DefaultTraits<Handle> {
+    static unsigned hash(Handle const& h) { return Traits<HANDLE>::hash(h.handle); }
+};
+template<>
+constexpr bool IsHashCompatible<HANDLE, Handle> = true;
+
+namespace Core {
+
+struct EventLoopTimer {
+    WeakPtr<EventReceiver> owner;
+    TimerShouldFireWhenNotVisible fire_when_not_visible = TimerShouldFireWhenNotVisible::No;
+};
+
+struct ThreadData {
+    static ThreadData& the()
+    {
+        thread_local OwnPtr<ThreadData> thread_data = make<ThreadData>();
+        return *thread_data;
+    }
+
+    ThreadData()
+    {
+        wake_event.handle = CreateEvent(NULL, FALSE, FALSE, NULL);
+        VERIFY(wake_event.handle);
+    }
+
+    // Each thread has its own timers, notifiers and a wake event.
+    HashMap<Handle, EventLoopTimer> timers;
+    HashMap<Handle, Notifier*> notifiers;
+
+    // The wake event is used to notify another event loop that someone has called wake().
+    Handle wake_event;
+};
+
+EventLoopImplementationWindows::EventLoopImplementationWindows()
+    : m_wake_event(ThreadData::the().wake_event.handle)
+{
+}
+
+int EventLoopImplementationWindows::exec()
+{
+    for (;;) {
+        if (m_exit_requested)
+            return m_exit_code;
+        pump(PumpMode::WaitForEvents);
+    }
+    VERIFY_NOT_REACHED();
+}
+
+size_t EventLoopImplementationWindows::pump(PumpMode)
+{
+    auto& thread_data = ThreadData::the();
+    auto& notifiers = thread_data.notifiers;
+    auto& timers = thread_data.timers;
+
+    size_t event_count = 1 + notifiers.size() + timers.size();
+    VERIFY(event_count <= MAXIMUM_WAIT_OBJECTS);
```

Comment:
This seems likely to trip on 'interesting' web pages. A fixme suggesting one of the workarounds from the docs might be appropriate.

> To wait on more than MAXIMUM_WAIT_OBJECTS handles, use one of the following methods:
>   - Create a thread to wait on MAXIMUM_WAIT_OBJECTS handles, then wait on that thread plus the other handles. Use this technique to break the handles into groups of MAXIMUM_WAIT_OBJECTS.
>  - Call [RegisterWaitForSingleObject](https://learn.microsoft.com/en-us/windows/desktop/api/winbase/nf-winbase-registerwaitforsingleobject) or [SetThreadpoolWait](https://learn.microsoft.com/en-us/windows/win32/api/threadpoolapiset/nf-threadpoolapiset-setthreadpoolwait) to wait on each handle. The thread pool waits efficiently on the handles and assigns a worker thread after the object is signaled or the time-out interval expires.